File: 00781. Rabbits in Forest/solution.cs
/**
 * Problem: 781. Rabbits in Forest
 *
 * Difficulty: Medium
 *
 * Language: C#
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

public class Solution
{
  public int NumRabbits(int[] answers)
  {
    // Frequency array for answers 0-1000 (stack allocated for performance)
    Span<int> answerGroups = stackalloc int[1001];

    // Count occurrences of each answer
    foreach (int answer in answers)
      answerGroups[answer]++;

    // Initialize total rabbits count
    int totalRabbits = 0;

    // Process answers from 1 to 1000 (answer 0 handled separately)
    for (int i = 1; i < answerGroups.Length; i++)
    {
      // Skip if no rabbits gave this answer
      if (answerGroups[i] == 0)
        continue;

      // Calculate number of groups needed for this answer
      // Each group has (i + 1) rabbits (i same-color rabbits + 1)
      // Formula: ceil(groups[i] / (i + 1)) * (i + 1)
      totalRabbits += ((answerGroups[i] + i) / (i + 1)) * (i + 1);
    }

    // Add rabbits that answered 0 (they are unique, each is a separate color)
    return totalRabbits + answerGroups[0];
  }
}

File: 00781. Rabbits in Forest/solution.ts
/**
 * Problem: 781. Rabbits in Forest
 *
 * Difficulty: Medium
 *
 * Language: TypeScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Minimum number of rabbits in the forest
 *
 * @param answers - Answers from rabbits
 *
 * @returns Minimum number of rabbits
 */
const numRabbits = (answers: number[]): number => {
  // Store the number of rabbits for each answer
  const frequencyMap = new Map<number, number>()

  // Store the minimum number of rabbits
  let totalRabbits = 0

  // Iterate through the answers
  for (let i = 0; i < answers.length; i++) {
    // Current answer
    const currentAnswer = answers[i]

    // If the current answer is not in the map or the count is 0
    if (
      !frequencyMap.has(currentAnswer) ||
      frequencyMap.get(currentAnswer) === 0
    ) {
      totalRabbits += currentAnswer + 1
      frequencyMap.set(currentAnswer, currentAnswer)
    }
    // Decrement the count for the current answer
    else frequencyMap.set(currentAnswer, frequencyMap.get(currentAnswer)! - 1)
  }

  // Return the minimum number of rabbits
  return totalRabbits
}
